File: runtime/CSlib/binary_port.cs
/*=====================================================================*/
/*    /users2GB/serrano2/bigloo/runtime/CSlib/binary_port.cs           */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Dec  3 11:33:29 2000                          */
/*    Last change :  Thu Feb  1 19:35:05 2007 (serrano)                */
/*    Copyright   :  2000-07 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    The binary port JVM connection                                   */
/*=====================================================================*/
using System;
using System.IO;

namespace bigloo 
{
   /*---------------------------------------------------------------------*/
   /*    BINARY_PORT ...                                                  */
   /*---------------------------------------------------------------------*/
   public sealed class binary_port: obj 
   {
      private static readonly byte[] MAGIC_WORD= foreign.getbytes( "1966" );
      public readonly Stream stream;

      public binary_port( Stream  stream ) 
	 {
	    this.stream= stream;
	 }

      public binary_port close() 
	 {
	    stream.Close();
	    return this;
	 }

      public binary_port flush() 
	 {
	    stream.Flush();
	    return this;
	 }

      public override void write( output_port  p ) 
	 {
	    p.write( "#<binary:port:" );
	    p.write( stream.ToString() );
	    p.write( ">" );
	 }

      public obj output_obj( Object  obj )
	 {

	    // the magic key
	    byte[] _string= foreign.__cb__.obj_to_string( obj );

	    stream.Write( MAGIC_WORD, 0, 4 );

	    // the object length
	    byte[] slen= new byte[4];
	    int clen= _string.Length;

	    slen[0]= (byte)clen;
	    slen[1]= (byte)(clen >> 8);
	    slen[2]= (byte)(clen >> 16);
	    slen[3]= (byte)(clen >> 24);

	    stream.Write( slen, 0, 4 );

	    // the serialized object
	    stream.Write( _string, 0, clen );

	    return this;
	 }

      public Object input_obj() 
	 {
	    // magic key
	    byte[] magic= new byte[4];
	    int size= stream.Read( magic, 0, 4 );

	    if (size == -1)
	       return eof._eof;

	    if ( (size != 4)
		 || (magic[0] != MAGIC_WORD[0])
		 || (magic[1] != MAGIC_WORD[1])
		 || (magic[2] != MAGIC_WORD[2])
		 || (magic[3] != MAGIC_WORD[3]))
	       foreign.fail( "input_obj", "corrupted file", this );

	    // the object length
	    byte[] slen= new byte[4];

	    size= stream.Read( slen, 0, 4 );
	    if (size != 4)
	       foreign.fail( "input_obj", "corrupted file", this );

	    int clen= (slen[0] & 0xff)
	       + ((slen[1] & 0xff) << 8)
	       + ((slen[2] & 0xff) << 16)
	       + ((slen[3] & 0xff) << 24);

	    // we start reading
	    byte[] _string= new byte[clen];

	    stream.Read( _string, 0, clen );
      
	    // unserialize the object
	    return foreign.__cb__.string_to_obj( _string );
	 }
   }
}

File: runtime/Jlib/binary_port.java
/*=====================================================================*/
/*    serrano/prgm/project/bigloo/runtime/Jlib/binary_port.java        */
/*    -------------------------------------------------------------    */
/*    Author      :  Manuel Serrano                                    */
/*    Creation    :  Sun Dec  3 11:33:29 2000                          */
/*    Last change :  Thu Jul 16 07:57:20 2015 (serrano)                */
/*    Copyright   :  2000-15 Manuel Serrano                            */
/*    -------------------------------------------------------------    */
/*    The binary port JVM connection                                   */
/*=====================================================================*/
package bigloo;

import java.io.*;

/*---------------------------------------------------------------------*/
/*    BINARY_PORT ...                                                  */
/*---------------------------------------------------------------------*/
public class binary_port extends obj {
   private static final byte[] MAGIC_WORD= "1966".getBytes();
   public final Object stream;

   public binary_port( final InputStream stream ) {
      this.stream = stream;
   }

   public binary_port( final OutputStream stream ) {
      this.stream = stream;
   }

   public binary_port close() throws IOException  {
      if( stream instanceof InputStream )
	 ((InputStream)stream).close();
      else
	 if( stream instanceof OutputStream )
	    ((OutputStream)stream).close();

      return this;
   }

   public obj flush() throws IOException {
      if( stream instanceof OutputStream )
	 ((OutputStream)stream).flush();

      return this;
   }

   public void write( final output_port p ) {
      p.write( "#<binary:" );

      if( stream instanceof InputStream )
	 p.write( "input_port:" );
      else
	 p.write( "output_port:" );

      p.write( stream.toString() );
      p.write( ">" );
   }

   public obj output_obj( final Object obj ) throws IOException {
      final OutputStream file = (OutputStream)stream;

      /* the magic key */
      final byte[] string = bigloo.runtime.Unsafe.intext.obj_to_string( obj, bigloo.foreign.BFALSE );

      file.write( MAGIC_WORD );

      /* the object length */
      final byte[] slen = new byte[4];
      final int clen = string.length;

      slen[0] = (byte)clen;
      slen[1] = (byte)(clen >> 8);
      slen[2] = (byte)(clen >> 16);
      slen[3] = (byte)(clen >> 24);

      file.write( slen );

      /* the serialized object */
      file.write( string );

      return this;
   }

   public Object input_obj() throws IOException {
	 final InputStream file = (InputStream)stream;

	 /* magic key */
	 final byte[] magic = new byte[4];
	 int size = file.read( magic );

	 if( size == -1 )
	    return eof.eof;

	 if( (size != 4)
	     || (magic[0] != MAGIC_WORD[0])
	     || (magic[1] != MAGIC_WORD[1])
	     || (magic[2] != MAGIC_WORD[2])
	     || (magic[3] != MAGIC_WORD[3]) )
	    foreign.fail( "input_obj", "corrupted file", this );

	 /* the object length */
	 final byte[] slen = new byte[4];

	 size = file.read( slen );
	 if( size != 4 )
	    foreign.fail( "input_obj", "corrupted file", this );

	 final int clen = (slen[0] & 0xff)
	    + ((slen[1] & 0xff) << 8)
	    + ((slen[2] & 0xff) << 16)
	    + ((slen[3] & 0xff) << 24);

	 /* we start reading */
	 final byte[] string = new byte[clen];

	 file.read( string );

	 /* unserialize the object */
	 return bigloo.runtime.Unsafe.intext.string_to_obj( string, false, false );
      }
}

